### PR TITLE
new lint: match_like_matches_macro

### PR DESCRIPTION
**Pull request by robojumper**
Suggests using the `matches!` macro from `std` where appropriate.

`redundant_pattern_matching` has been moved into the `matches` pass to allow suppressing the suggestion where `is_some` and friends are a better replacement.

changelog: new lint: `match_like_matches_macro`

### CODE REVIEW

**Comment by camsteffen** [comment]
Should this also include cases like the following?

```rust
match o {
    Some(1) => 'y',
    _ => 'n',
}
```
to
```rust
if matches!(o, Some(1)) { 'y' } else { 'n' }
```

**Comment by robojumper** [comment]
@camsteffen The transformation you describe first replaces `match` with `if let` (which is a more aggressive form of the currently already *pedantic-allow* [`single_match_else`](https://rust-lang.github.io/rust-clippy/master/index.html#single_match_else) lint) and then replaces the `if let` header with `if matches!`.

I think that transformation should not be part of this lint, as the `matches!` macro specifically shines when it allows the match arms / block bodies (only containing only `true` and `false`) to be omitted and replaces the entire expression. That's also why I suggested this lint be in the `style` group and warn by default.

File: clippy_lints/src/assign_ops.rs [hunk 1]
```diff
@@ -237,6 +237,7 @@ fn is_commutative(op: hir::BinOpKind) -> bool {
     use rustc_hir::BinOpKind::{
         Add, And, BitAnd, BitOr, BitXor, Div, Eq, Ge, Gt, Le, Lt, Mul, Ne, Or, Rem, Shl, Shr, Sub,
     };
+    #[allow(clippy::match_like_matches_macro)]
```
Comment (phansch) [inline clippy_lints/src/assign_ops.rs:240]: Would it make sense to only lint if the second arm matches on `PatKind::Wild`? That would make the lint a bit less noisy and prevent this FP, I think?

Comment (robojumper) [reply clippy_lints/src/assign_ops.rs:240]: I had initially considered that. However, that would mean a match with the arms `{Some(_), None}` would not be caught while `{Some(_), _}` would be despite them not being particularly different. I only hit one possibly intentionally exhaustive match in the entire clippy repository, and that `allow` serves as a (somewhat obscure) "hey, this is intentionally exhaustive".

Comment (robojumper) [reply clippy_lints/src/assign_ops.rs:240]: On second thought, `{Some(_), None}` will hit the `redundant_pattern_matching` lint anyway. It probably makes sense to restrict this lint to non-exhaustive matches.

**Comment by phansch** [comment]
LGTM, thanks!

@bors r+

**Pull request merged**